File: csharp/csharp/bob/Bob.cs
﻿using System;
using System.Linq;

namespace Exercises.csharp.bob
{
	internal class Bob
	{
		public string Hey(string input)
		{
			if (String.IsNullOrWhiteSpace(input))
				return "Fine. Be that way!";

			// The input is 'yelling' if it contains letters and every letter is capitalized
			if (input.Any(char.IsLetter) && input.All(c => !char.IsLetter(c) || char.IsUpper(c)))
				return "Whoa, chill out!";

			if (input.EndsWith("?"))
				return "Sure.";

			return "Whatever.";
		}
	}
}

File: go/bob/bob.go
// Package bob contains functions related to Bob the teenager
package bob

import (
	"strings"
	"unicode"
)

// Hey returns Bob's response to the given remark
func Hey(remark string) string {
	trimmed := strings.TrimSpace(remark)
	if len(trimmed) == 0 {
		return "Fine. Be that way!"
	}

	isQuestion := strings.HasSuffix(trimmed, "?")
	hasLetters := false
	allCaps := true

	for _, char := range trimmed {
		if unicode.IsLetter(char) {
			hasLetters = true
			if unicode.IsLower(char) {
				allCaps = false
			}
		}
	}

	if hasLetters && allCaps && isQuestion {
		return "Calm down, I know what I'm doing!"
	} else if hasLetters && allCaps {
		return "Whoa, chill out!"
	} else if isQuestion {
		return "Sure."
	} else {
		return "Whatever."
	}
}
